sqlglot/parser.py

```diff
@@ -6433,14 +6436,36 @@ def _match_text_seq(self, *texts, advance=True):
         return True
 
     def _replace_lambda(
-        self, node: t.Optional[exp.Expression], lambda_variables: t.Set[str]
+        self, node: t.Optional[exp.Expression], expressions: t.List[exp.Expression]
     ) -> t.Optional[exp.Expression]:
         if not node:
             return node
 
+        lambda_types = {}
+
+        for this in expressions:
+            if isinstance(this, exp.Cast):
+                typ = this.to
+                this = this.this
+            else:
+                typ = None
+
+            lambda_types[this.name] = typ
+
         for column in node.find_all(exp.Column):
-            if column.parts[0].name in lambda_variables:
+            name = column.parts[0].name
+            if name in lambda_types:
                 dot_or_id = column.to_dot() if column.table else column.this
+
+                typ = lambda_types.get(name)
```

Comment:
```suggestion
            typ = lambda_types.get(column.parts[0].name)
            if typ is not None:
                dot_or_id = column.to_dot() if column.table else column.this
```

Reply:
are you sure this is right? we need to convert to a dot even if there is no typ